### ISSUE TITLE
How to handle a list of strings? Can we parse dates as moment?

### ISSUE THREAD

**Issue by boy51** [davidmdm/myzod#15]
Hi, awesome repo! Having a good time using it. I have two questions, hope its ok to post here.

1. How would I handle a type that is not a string but a list of accepted strings?
```status:  "active" | "disabled"```
I'm not sure if `myzod.enum()` is the way to go for this? I'm using `myzod.string()` and regex to test my values, but a downside is that you don't get typescript representation.

2. I'm not sure if there is a feature like this. But since we have `myzod.date()`, is there a way to use a different date constructor such as `moment`? Right now I'm parsing my dates as dates and manually make them moments. This is highly inconvenient as I'm forced to have seperate types for my axios response and my actual state because I don't want to broaden my state interface.

Keep up the work :)

**Comment by davidmdm** [comment]
Hi!

Yes it is possible to have typed literals with myzod. Check out the doc on `myzod.literal(s)`

you could have a schema like this:

```typescript
import * as z from 'myzod';
const statusSchema = z.literals('active', 'disabled');
type Status = z.Infer<typeof statusSchema>; // -> 'active' | 'disabled' 
```

As for the interoperability with moment.js types I have no direct solution within `myzod` at this time, and hadn't thought of it. I am thinking about what kind of api could satisfy this. I am thinking of maybe a `map` method...

```typescript
const dateSchema = z.date().map(value => moment(value)); // -> Type<Moment>
type Date = Infer<typeof dateSchema>; // -> Moment
```

I will need to think on it and make sure this is a change I want to make but that could be a possible solution!

**Comment by boy51** [comment]
Thanks, seems I have missed the literals method.

It would be a great feature if myzod could give me typed moments. Maybe it can be done like in `@material-ui/pickers`? I can see that this is a tricky thing to do. But it would be awesome :)

EDIT: Obviously you could support other libraries like date-fns in that way, although I imagine it shouldn't be too hard for moment as moment is just a wrapper for a regular date. I'm not sure how the other libraries work.

**Comment by davidmdm** [comment]
@boy51 

I am not familiar with how `material-ui/pickers` solves this problem? Can you show me a quick example of the api and how it would work here?

How do you feel about the `map` method I proposed above? It would allow you to map and type to a different type ie Date -> Moment or Date -> Date-FNS

**Comment by boy51** [comment]
Sure, I'm talking about [something like this pattern](https://material-ui-pickers.dev/getting-started/installation). I just think it would be nice to be able to provide a date library. Maybe its a bad example because they are mainly dealing with the DOM of course.

Map is a nice and simple solution as it would be able to return the proper types. I imagine adding different date constructors to the Infer system would be much trickier if not impossible.

Actually I think your proposal is quite sexy, but the question is whether it will pass in the date or the string as value. I would rather have the original string as dates can get messy. And actually this would be really amazing as I am converting timezones all the time so I could immediately do stuff like `z.date().map(value => moment.utc(value).tz(timezone))`.

Would I be using `z.string().map()` or `z.date().map()`?

**Comment by davidmdm** [comment]
Yeah the issue with material-ui of providing an adaptor for each library is that it never ends. Letting the user validate a base JavaScript type and then map it to what they want takes the pressure away from me as the author of the lib and allows the user to do anything they want.

In this case you could do either string().map() or z.date().map() but I would prefer the `date` approach simply because myzod will validate that the value was a date string and so your `moment` should not be invalid. Unless of course you rather moment validates what is an invalidate date. Then you could just throw an error from the mapping function or something...

Anyways I am satisfied with this conversation. I think I will start developing this `map` feature today. I'll close this issue once it is released.

**Comment by boy51** [comment]
>[...] simply because myzod will validate that the value was a date string and so your `moment` should not be invalid. Unless of course you rather moment validates what is an invalidate date. Then you could just throw an error from the mapping function or something...

The problem is more about consistency. I'm dealing with all kind of different dates and going from string to date to moment can mess up your dates. I.e. if you have a date string with ambiguous timezone, passing it in a date constructor will actually adjust it to the local timezone, and then moment will not have an ambiguous input but an input with the local timezone, and that could be incorrect.

That being said verifying the date is useful none the less so maybe there could be a second param that has the original value or something?

**Comment by davidmdm** [comment]
Hi @boy51 ,

Perhaps this is too little too late but I have released the `map` feature in v1.2.0.

Here is the example from the docs using ObjectId but it would work for a Date type or Moment too:
```typescript
const ObjectIDSchema = myzod
                         .string()
                         .withPredicate(ObjectId.isValid, 'must be an object ID')
                         .map(value => new ObjectId(value));

// Infer<ObjectIDSchema> === ObjectId

const id = ObjectIDSchema.parse('507c7f79bcf86cd7994f6c0e');
id instanceof ObjectId; // true

const id2 = ObjectIDSchema.parse('some string'); // Throws VaidationError with message "must be an object ID"'
```

Closing your issue now but would love feedback if this helps.
thx.
